library/psa_crypto.c

```diff
@@ -6360,4 +6373,583 @@ psa_status_t psa_crypto_init( void )
+
+    if( type != PSA_KEY_TYPE_PASSWORD &&
+        type != PSA_KEY_TYPE_PASSWORD_HASH )
+    {
+        status = PSA_ERROR_INVALID_ARGUMENT;
+        goto error;
+    }
+
+    operation->data.inputs.password = mbedtls_calloc( 1, slot->key.bytes );
+    if( operation->data.inputs.password == NULL )
+    {
+        status = PSA_ERROR_INSUFFICIENT_MEMORY;
+        goto error;
+    }
+
+    memcpy( operation->data.inputs.password, slot->key.data, slot->key.bytes );
+    operation->data.inputs.password_len = slot->key.bytes;
+    operation->data.inputs.key_lifetime = attributes.core.lifetime;
+error:
+    unlock_status = psa_unlock_key_slot( slot );
+    return( ( status == PSA_SUCCESS ) ? unlock_status : status );
+}
+
+psa_status_t psa_pake_set_user(
+    psa_pake_operation_t *operation,
+    const uint8_t *user_id,
+    size_t user_id_len )
+{
+    (void)user_id;
+
+    if( operation->stage != PSA_PAKE_OPERATION_STAGE_COLLECT_INPUTS )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( operation->alg == PSA_ALG_NONE )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( user_id_len == 0 )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    return( PSA_ERROR_NOT_SUPPORTED );
+}
+
+psa_status_t psa_pake_set_peer(
+    psa_pake_operation_t *operation,
+    const uint8_t *peer_id,
+    size_t peer_id_len )
+{
+    (void)peer_id;
+
+    if( operation->stage != PSA_PAKE_OPERATION_STAGE_COLLECT_INPUTS )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( operation->alg == PSA_ALG_NONE )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( peer_id_len == 0 )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    return( PSA_ERROR_NOT_SUPPORTED );
+}
+
+psa_status_t psa_pake_set_role(
+    psa_pake_operation_t *operation,
+    psa_pake_role_t role )
+{
+    if( operation->stage != PSA_PAKE_OPERATION_STAGE_COLLECT_INPUTS )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( operation->alg == PSA_ALG_NONE )
+        return( PSA_ERROR_BAD_STATE );
+
+    if( role != PSA_PAKE_ROLE_NONE &&
+        role != PSA_PAKE_ROLE_FIRST &&
+        role != PSA_PAKE_ROLE_SECOND &&
+        role != PSA_PAKE_ROLE_CLIENT &&
+        role != PSA_PAKE_ROLE_SERVER )
+    {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    operation->data.inputs.role = role;
+
+    return( PSA_SUCCESS );
+}
+
+static psa_status_t psa_pake_complete_inputs(
+    psa_pake_operation_t *operation )
+{
+    psa_jpake_computation_stage_t *computation_stage =
+        &operation->computation_stage.data.jpake_computation_stage;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    uint8_t* password = operation->data.inputs.password;
+    size_t password_len = operation->data.inputs.password_len;
+
+    if( operation->alg == PSA_ALG_NONE ||
+        operation->data.inputs.password_len == 0 ||
+        operation->data.inputs.role == PSA_PAKE_ROLE_NONE )
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    status = psa_driver_wrapper_pake_setup( operation,
```

Comment:
There's a memory management problem here: `psa_driver_wrapper_pake_setup` expects `operation->data.ctx` to be the active member of the union, but the call here expects `operation->data.inputs` to be the active member. `inputs` is live until the driver's setup entry point returns, and `ctx` is live from the point the driver's setup entry point is called (or rather, just before: we should zero it out).

This function needs to make a stack copy of one of the two structures (the one you expect to be smaller).

Reply:
The inputs are only 20 bytes, so should be ok to make a copy on stack. This will solve the problem with two pointer arguments in driver setup entry point ( `inputs` and driver `context`) that points to the same memory.

So the `psa_driver_wrapper_pake_setup` should do the following actions:
1. make a local copy of `inputs` structure.
2. perform driver dispatch and select active context.
3. clear active context.
4. call setup driver entry point passing pointer to active context and pointer to local copy of inputs.

Reply:
I recommend doing it in `psa_pake_complete_inputs`, not in the driver wrapper. The driver wrapper is generated code, and the less complex it is, the better. It should only have the minimum logic needed to choose which driver to dispatch and call the appropriate function.